### PR TITLE
chore: unify sourcery execution [WPB-9377]

### PR DESCRIPTION
**Pull request by markusfassbender**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9377" title="WPB-9377" target="_blank">TaskWPB-9377</a>  [iOS] unify sourcery code generation for swift packages
  </td></table>
  <br />
 

### Issue

After discussion in the last API PR we like to find a way to unify the execution of [`Sourcery`](https://github.com/krzysztofzablocki/Sourcery):
https://github.com/wireapp/wire-ios/pull/1445#discussion_r1603114496

Important requirements
- Keep low build times, so we prefer a binary over the code with many sub-dependencies
- Keep the automated execution for these target, that are not Swift Packages
- Offer a manual way to execute mock generation for Swift Packages, better for all targets.

### Testing

- Locally run `TeamRepositoryTests` that uses `WireAPISupport`
- Build the project
- Run the overall test suite https://github.com/wireapp/wire-ios/actions/runs/9189136860

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

### CODE REVIEW

File: WireAPI/Sourcery/config.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+sources:
+- ../Sources/WireAPI
+templates:
+- ./AutoMockable.stencil
```
Comment (markusfassbender) [inline WireAPI/Sourcery/config.yml]: We could also unify the stencil over all targets (and just make unique ones if something differs)?

File: wire-ios/Wire-iOS.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -9057,6 +9075,24 @@
...
+			);
+			inputPaths = (
+			);
+			outputFileListPaths = (
+			);
+			outputPaths = (
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "# Wire-iOS\n../scripts/run-sourcery.sh --config ./Tests/Sourcery/config.yml\n\n# WireAPI\n../scripts/run-sourcery.sh --config ../WireAPI/Sourcery/config.yml\n";
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS.xcodeproj/project.pbxproj]: Of course `WireDataModel` etc. are missing currently, as it should be only one example.

File: WireAPI/Sourcery/config.yml (hunk 1 above)
Comment (johnxnguyen) [reply WireAPI/Sourcery/config.yml]: Yeah that would be great.

**Review by johnxnguyen — COMMENTED** [review]
For me it looks good, maybe you could show it in action in the iOS lounge?

File: WireAPI/Sourcery/config.yml (hunk 1 above)
Comment (caldrian) [reply WireAPI/Sourcery/config.yml]: At least the `import`s differ.

Comment (markusfassbender) [reply WireAPI/Sourcery/config.yml]: @caldrian  I don't understand what you mean about the imports? I can not spot hardcoded imports in the stencil files.

File: Packages/SourceryPlugin/.gitignore [hunk 3]
```diff
@@ -0,0 +1,7 @@
+.DS_Store
+/.build
+/Packages
+xcuserdata/
+DerivedData/
+.swiftpm/configuration/registries.json
+.swiftpm/xcode/package.xcworkspace/contents.xcworkspacedata
```
Comment (netbe) [inline Packages/SourceryPlugin/.gitignore]: what are those files ? and why do we ignore them?

File: WireAPI/Package.swift [hunk 4]
```diff
@@ -5,7 +5,8 @@ import PackageDescription
 let package = Package(
     name: "WireAPI",
     platforms: [
-        .iOS(.v15)
+        .iOS(.v15),
+        .macOS(.v12)
```
Comment (markusfassbender) [inline WireAPI/Package.swift:9]: This was a quick solution for the issue from App Store Connect this morning and discussed with @netbe and @johnxnguyen :
Screenshot 2024-05-22 at 11 53 31

File: WireAPI/Sourcery/config.yml (hunk 1 above)
Comment (caldrian) [reply WireAPI/Sourcery/config.yml]: Sorry, I thought you mean the stencils.

File: wire-ios-request-strategy/WireRequestStrategy.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -3170,6 +3171,10 @@
 /* End XCConfigurationList section */
 
 /* Begin XCLocalSwiftPackageReference section */
+		E67F36282BFDE35000B83E0F /* XCLocalSwiftPackageReference "../Packages/SourceryPlugin" */ = {
```
Comment (caldrian) [inline wire-ios-request-strategy/WireRequestStrategy.xcodeproj/project.pbxproj]: Aren't these changes which you just made for demo purposes in our call?

File: WireAPI/Sources/WireAPISupport/Mocks.swift [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+/// This is just a placeholder to give the compiler some food.
+/// To auto generate mocks via 'sourcery' it uses the plugin configured in `Package.swift`.
+/// The generated mocks are processed from the sandbox directory and are not visible in the project folder:
+/// https://github.com/apple/swift-package-manager/blob/main/Documentation/Plugins.md#implementing-the-build-tool-plugin-script
+enum Mocks { }
+
+extension Mocks {
+
+    /// An Example how generated mocks are structured.
+    public class MockExample {
```
Comment (caldrian) [inline WireAPI/Sources/WireAPISupport/Mocks.swift]: I would just remove it.
Xcode shows you the signature of methods and properties.
Also, all this is defined in the stencil files.

File: SourceryPlugin/Plugins/SourceryPlugin/SourceryPlugin.swift [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+        guard let configuration else {
+            Diagnostics.error(
+"""
+No configurations found for target \(target.name). If you would like to generate sources for this \
+target include a `\(Constant.configFileName)` either in:
+1. root of package
+2. target directory
+3. target directory subfolder named 'Sourcery'
+"""
```
Comment (caldrian) [inline SourceryPlugin/Plugins/SourceryPlugin/SourceryPlugin.swift]:
```suggestion
            Diagnostics.error("""
                No configurations found for target \(target.name). If you would like to generate sources for this \
                target include a `\(Constant.configFileName)` either in:
                1. root of package
                2. target directory
                3. target directory subfolder named 'Sourcery'
                """
```

File: SourceryPlugin/Plugins/SourceryPlugin/SourceryPlugin.swift [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+import Foundation
+import PackagePlugin
+
+@main
+struct SourceryPlugin {
+
+    private enum Constant {
+        enum Arguments {
+            static let config = "config"
```
Comment (caldrian) [inline SourceryPlugin/Plugins/SourceryPlugin/SourceryPlugin.swift]: These values are defined by Sourcery, we'll never be able to use any different value here.
Hence, I don't see the point of defining constants.

**Review by caldrian — APPROVED** [review]

File: SourceryPlugin/README.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+The plugin's environment provides three paths that can be used in the configuration:
+1. `DERIVED_SOURCES_DIR`: The working directory in the sandbox that contains the results.
+2. `PACKAGE_ROOT_DIR`: The package root directory.
+3. `TARGET_DIR`: The target directory.
+
+
+## Future ideas:
+- Create a shared GitHub repository for `SourceryPlugin` instead of this local implementation.
+- Implement `XcodePluginContext` to run the plugin as a "Run Build Tool Plug-ins" in other targets that are not Swift Packages, like WireDataModel.
+- Share the `AutoMockable.stencil` file between all targets or deliver defaults from the plugin. Apple's documentation suggests that processing resources like assets might become easier in the future.
```
Comment (netbe) [inline SourceryPlugin/README.md:46]: we can already create a ticket for that one

File: WireAPI/Sources/WireAPISupport/Mocks.swift (hunk 6 above)
Comment (netbe) [reply WireAPI/Sources/WireAPISupport/Mocks.swift]: I get @caldrian point. The example was more for someone that onboard on the project.
I would add a topic section in Wire-iOS docC that mention we use Sourcery for mocks.

Comment (markusfassbender) [reply WireAPI/Sources/WireAPISupport/Mocks.swift]: Done

File: Packages/SourceryPlugin/.gitignore (hunk 3 above)
Comment (markusfassbender) [reply Packages/SourceryPlugin/.gitignore]: Honestly I just copied it from `WireAPI` package, but improved it slightly now to ignore the whole folder `.swiftpm/xcode/`.

File: SourceryPlugin/README.md (hunk 9 above)
Comment (markusfassbender) [reply SourceryPlugin/README.md:46]: We could, but on the other hand we have a lot of tickets that are waiting in pipeline and the backlog is huge. I would rather create the tickets if we really want to work on this soon, ok?

File: wire-ios-request-strategy/WireRequestStrategy.xcodeproj/project.pbxproj (hunk 5 above)
Comment (markusfassbender) [reply wire-ios-request-strategy/WireRequestStrategy.xcodeproj/project.pbxproj]: Right, thanks! removed it with 1686dc59efa741ea1aac79cbd63012a4bf7e3da4

File: SourceryPlugin/Plugins/SourceryPlugin/SourceryPlugin.swift (hunk 8 above)
Comment (markusfassbender) [reply SourceryPlugin/Plugins/SourceryPlugin/SourceryPlugin.swift]: Done

**Review by netbe — APPROVED** [review]
LGTM;)

**Pull request merged**